contrib/rabbitmq/pkg/adapter/adapter.go

```diff
@@ -0,0 +1,237 @@
+	RoutingKey       string
+	Durable          bool
+	DeleteWhenUnused bool
+	Exclusive        bool
+	NoWait           bool
+}
+
+type Adapter struct {
+	Brokers        string
+	Topic          string
+	ExchangeConfig ExchangeConfig
+	QueueConfig    QueueConfig
+	SinkURI        string
+	client         client.Client
+}
+
+func (a *Adapter) initClient() error {
+	if a.client == nil {
+		var err error
+		if a.client, err = kncloudevents.NewDefaultClient(a.SinkURI); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (a *Adapter) Start(ctx context.Context, stopCh <-chan struct{}) error {
+	logger := logging.FromContext(ctx)
+
+	if err := a.initClient(); err != nil {
+		logger.Error("Failed to create cloudevent client", zap.Error(err))
+		return err
+	}
+
+	logger.Info("Starting with config: ", zap.Any("adapter", a))
+
+	conn, err := amqp.Dial(a.Brokers)
+	if err != nil {
+		logger.Error(err)
+	}
+	defer conn.Close()
+
+	ch, err := conn.Channel()
+	if err != nil {
+		logger.Error(err)
+	}
+	defer ch.Close()
+
+	queue, err := a.StartAmqpClient(ctx, &ch)
+
+	return a.pollForMessages(ctx, &ch, queue, stopCh)
+}
+
+func (a *Adapter) StartAmqpClient(ctx context.Context, ch *wabbit.Channel) (*wabbit.Queue, error) {
+	logger := logging.FromContext(ctx)
+	exchangeConfig := fillDefaultValuesForExchangeConfig(&a.ExchangeConfig, a.Topic)
+
+	err := (*ch).ExchangeDeclare(
+		exchangeConfig.Name,
+		exchangeConfig.TypeOf,
+		wabbit.Option{
+			"durable":  exchangeConfig.Durable,
+			"delete":   exchangeConfig.AutoDeleted,
+			"internal": exchangeConfig.Internal,
+			"noWait":   exchangeConfig.NoWait,
+		},)
+	if err != nil {
+		logger.Error(err)
+		return nil, err
+	}
+
+	queue, err := (*ch).QueueDeclare(
+		a.QueueConfig.Name,
+		wabbit.Option{
+			"durable":   a.QueueConfig.Durable,
+			"delete":    a.QueueConfig.DeleteWhenUnused,
+			"exclusive": a.QueueConfig.Exclusive,
+			"noWait":    a.QueueConfig.NoWait,
+		},)
+	if err != nil {
+		logger.Error(err)
+		return nil, err
+	}
+
+	routingKeys := strings.Split(a.QueueConfig.RoutingKey, ",")
+
+	for _, routingKey := range routingKeys {
+		err = (*ch).QueueBind(
+			queue.Name(),
+			routingKey,
+			exchangeConfig.Name,
+			wabbit.Option{
+				"noWait": a.QueueConfig.NoWait,
+			},)
+		if err != nil {
+			logger.Error(err)
+			return nil, err
+		}
+	}
+	return &queue, nil
+}
+
+func (a *Adapter) pollForMessages(ctx context.Context, channel *wabbit.Channel,
+	queue *wabbit.Queue, stopCh <-chan struct{}) error {
+	logger := logging.FromContext(ctx)
+
+	msgs, err := (*channel).Consume(
+		(*queue).Name(),
+		"",
+		wabbit.Option{
+			"noAck":     false,
+			"exclusive": a.QueueConfig.Exclusive,
+			"noLocal":   false,
+			"noWait":    a.QueueConfig.NoWait,
+		},)
+
+	if err != nil {
+		logger.Error(err)
+	}
+
+	for {
+		select {
+		case msg, ok := <-msgs:
+			if ok {
+				logger.Info("Received: ", zap.Any("value", string(msg.Body())))
+
+				if err := a.postMessage(ctx, msg); err != nil {
```

Comment:
What happens if the sink is overloaded?

Reply:
see and example for Apache Kafka, here:

https://github.com/knative/eventing-sources/pull/308/files

Only when message could be delivered, it's committed 

GCPUBSUB, does similar, `ack` or `nack` (if post failed)

Reply:
Hi @duglin, @matzew , thanks for pointing that out. I have added support for `ack` or `nack`.